src/query/api/v1/handler/prom/read.go

```diff
@@ -40,28 +41,84 @@ import (
 	"go.uber.org/zap"
 )
 
+// ReadHandlerHooks allows dynamic plugging into read request processing.
+type ReadHandlerHooks interface {
+	// OnParsedRequest gets invoked after parsing request arguments.
+	OnParsedRequest(
+		context.Context,
+		*http.Request,
+		models.RequestParams,
+	) (models.RequestParams, error)
+}
+
+// NewQueryFn creates a new promql Query.
+type NewQueryFn func(
+	engine *promql.Engine,
+	queryable promstorage.Queryable,
+	params models.RequestParams,
+) (promql.Query, error)
+
+var (
+	newRangeQueryFn = func(
+		engine *promql.Engine,
+		queryable promstorage.Queryable,
+		params models.RequestParams,
+	) (promql.Query, error) {
+		return engine.NewRangeQuery(
+			queryable,
+			params.Query,
+			params.Start,
+			params.End,
+			params.Step)
+	}
+
+	newInstantQueryFn = func(
+		engine *promql.Engine,
+		queryable promstorage.Queryable,
+		params models.RequestParams,
+	) (promql.Query, error) {
+		return engine.NewInstantQuery(
+			queryable,
+			params.Query,
+			params.Now)
+	}
```

Comment:
These could be a branch in `ServeHTTP`, e.g.:

```golang
var qry promql.Query
if h.instant {
  qry, err = engine.NewInstantQuery(/* ... */)
} else {
  qry, err = engine.NewRangeQuery(/* ... */)
}
```

Reply:
The branching and plumbing parts (DI) should be done in the root level of the code, here we have actual logic in place so by doing this we would just depend on implementation details of concrete queries and also violate DI.